### ISSUE TITLE
control state of individual taskflow in mapped task-group

### ISSUE THREAD

**Issue by potiuk** [apache/airflow#40543]
### Discussed in https://github.com/apache/airflow/discussions/33556

<div type='discussions-op-text'>

<sup>Originally posted by **ntnhaatj** August 20, 2023</sup>
### Description

Hi, as [my issue was raised here](https://github.com/apache/airflow/issues/25032#issuecomment-1684926654) which also included depth-first execution model context, I would like to open an issue to have more discussion on this matter.

### Use case/motivation

motivated from the depth-first execution model implemented in mapped TaskGroup:

```
                     ┌──────────────┐ ┌────────────────┐ ┌───────────┐
                 ┌──►│extract_file_1│►│transform_file_1│►│load_file_1│
                 │   └──────────────┘ └────────────────┘ └───────────┘
                 │
                 │   ┌──────────────┐ ┌────────────────┐ ┌───────────┐
 ┌─────────────┐ ├──►│extract_file_2│►│transform_file_2│►│load_file_2│
 │get_file_list├─┤   └──────────────┘ └────────────────┘ └───────────┘
 └─────────────┘ │
                 │         ...               ...              ...
                 │
                 │   ┌──────────────┐ ┌────────────────┐ ┌───────────┐
                 └──►│extract_file_N│►│transform_file_n│►│load_file_N│
                     └──────────────┘ └────────────────┘ └───────────┘
```

At present, the upstream / downstream list dependencies now only applied on DAGNode (which is the whole TaskGroup or Operator)

There might be better control over desired TaskFlow if we could enhance support for deeper upstream/downstream dependencies at the granularity of mapped task instances, instead of applying them to the entire TaskGroup.

In the model above, `extract_file_1` task downstream list should be only `transform_file_1` and `load_file_1` in order, rather than encompassing the whole mapped group `transform_file[]` and `load_file[]` as in the current implementation.

For instance, I scheduled my test DAG on `Airflow 2.7.0`:

```python
from airflow.decorators import dag, task_group, task
from airflow.operators.empty import EmptyOperator
from pendulum import datetime

files = ["a", "b", "c"]

@dag(start_date=datetime(2022, 12, 1), schedule=None, catchup=False)
def task_group_mapping_example():
    @task_group(group_id="etl")
    def etl_pipeline(file):
        e = EmptyOperator(task_id="e")
        t = EmptyOperator(task_id="t")
        l = EmptyOperator(task_id="l")

        e >> t >> l

    etl = etl_pipeline.expand(file=files)
    etl

task_group_mapping_example()
```

Clear etl.e[1] state with downstream will trigger all mapped task in `etl.t[]` and `etl.l[]` group.

<img width="877" alt="Screenshot 2023-08-19 at 18 46 42" src="https://github.com/apache/airflow/assets/35696768/dd110d7b-ecdf-4bcc-a466-a7d49b54ac87">

Thanks,

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)
</div>

**Comment by boraberke** [comment]
I would like to give this one a try. Could you please assign me?

**Comment by gavinhonl** [comment]
@potiuk - Thanks for raising this issue. I was about to create a new issue myself. IMO - the behaviour should be as follows:

- Select task in Grid (task_id[]) -> Clear task (with Downstream enabled): Clears all mapped tasks and their associated downstream tasks - Existing Behaviour
- Select mapped task (task_id[map_index]) -> Clear task (with Downstream enabled): Clears only the mapped task and associated downstream tasks - To-be Behaviour
- Select TaskGroup in Grid (taskgroup_id[]) -> Clear task (with Downstream enabled): Clears all tasks in the Taskgroup and associated downstream tasks - Existing Behaviour
-  Select TaskGroup in Grid (taskgroup_id[]) -> Clear task (with Downstream disabled): Clears all tasks in the Taskgroup - Existing Behaviour

**Comment by shahar1** [comment]
@boraberke it seems that @harjeevanmaan is currently working on this one.
If you're still interested, feel free to collaborate/review their PR (in #41384), or take over another issue (there are other open issues in the dynamic task mapping area)

**Comment by sharpnife** [comment]
@harjeevanmaan @shahar1  Any progress on this issue? 🙂

**Comment by shahar1** [comment]
> @harjeevanmaan @shahar1 Any progress on this issue? 🙂

Currently this issue is unassigned, so if it's important for you - you're welcome to join the efforts.

**Comment by potiuk** [comment]
Indeed. FYI. @sharpnife -> the best way to influence speed of implementing task is to roll sleeves up and implement it, rather than ask for startus. That's how open-source works. Generally the status is always as seen in the issue - if no-one works on it, well, no-one works on it.

**Comment by Shekharrajak** [comment]
In current implementation - when clearing etl.e[1] with downstream dependencies:

1. find_relevant_relatives() identifies etl.t and etl.l as downstream.
2. _get_relevant_map_indexes() should return map_index 1 for both (same mapped group).
3. Instead, it returns None or an incorrect value, causing all instances to be cleared.

I think we need a function to get the map_index for a relative task that's in the same mapped task group.

Please assign the task to me, let me share the PR.

**Comment by potiuk** [comment]
@Shekharrajak We are unassigning you from this issue as part of our updated [assignment policy](https://github.com/apache/airflow/blob/main/contributing-docs/04_how_to_contribute.rst#contribute-code-changes).

This is not meant to discourage your contribution — quite the opposite! You are still very welcome to work on this issue and submit a PR for it. Simply comment that you are working on it and open a PR when ready.

We found that formal assignments were not working well, as they often prevented others from contributing when the assignee was not actively working on the issue.

**Comment by adithya2754** [comment]
Hey @potiuk could you please assign me? I would like to give it a try

**Comment by shahar1** [comment]
> Hey [@potiuk](https://github.com/potiuk) could you please assign me? I would like to give it a try

We don't assign users to GitHub issues anymore by policy (with some exceptions that don't apply for this case).
You're welcome to work on it and create a Pull Request.

**Comment by gitedmond** [comment]
I tested the original scenario from this issue against current main (9f978d30097b030bb68f24d88e9e1e3da11e0c99), and I could not reproduce the reported behavior.
The regression test creates a serialized Dag with a three-item mapped TaskGroup containing extract >> transform >> load, creates a real DagRun, and verifies that all nine TaskInstance rows exist. It then calls the public clearTaskInstances API using the same mapped-task marker sent by the UI.
I tested indexes 0, 1, and 2 in both directions:
	•	Clearing etl.extract[index] with downstream tasks included returned and cleared only extract[index], transform[index], and load[index]. 
	•	Clearing etl.load[index] with upstream tasks included returned and cleared only extract[index], transform[index], and load[index]. 
	•	Preview mode left all states unchanged. 
	•	Actual clearing changed only the three TaskInstance rows associated with the selected index. 
All six parameterized cases passed against unchanged production code.
Since the original report used Airflow 2.7.0, the behavior appears to have been corrected by subsequent mapped-task handling changes, although I have not identified the exact fixing commit. This result does not imply that the original report was user error.
Could a maintainer confirm whether this issue can be closed as already fixed on current main? I can also provide the regression test if retaining explicit coverage for this behavior would be useful.
